### DISCUSSION TITLE
Interface kernel with coupled variable

### DISCUSSION THREAD

**Discussion by Chaoyue-Liu** [idaholab/moose#22366] [Q&A General] [answered]
I am building a customized interface kernel with coupled variable involved, but I encountered problem that the interface kernel cannot be calculated. Could you help me to check where I did wrong in the code? Thank you!

The model is like the following:
![image](https://user-images.githubusercontent.com/95432912/195098113-c1db1277-2e62-4ad7-8446-4ab5605a897f.png)

My interface kernels are the followings, BV is for equation 1), BVFLUX is for equation 2).
BV:
```
#include "BV.h"

registerMooseObject("batteryApp", BV);

InputParameters
BV::validParams()
{
  InputParameters params = ADInterfaceKernel::validParams();
  params.addParam<MaterialPropertyName>("i0","name of i0");
  params.addParam<MaterialPropertyName>("Eeq","name of Eeq");
  params.addCoupledVar("cl","solid concentration");
  params.addClassDescription("Implements a reaction to establish ReactionRate=k_f*u-k_b*v "
                             "at interface.");
  return params;
}

BV::BV(const InputParameters & parameters)
  : ADInterfaceKernel(parameters),
  _lstandard(getMaterialProperty<Real>("i0")),
  _Eeq(getMaterialProperty<Real>("Eeq")),
  _cl(coupledValue("cl"))
{
}

ADReal
BV::computeQpResidual(Moose::DGResidualType type)
{
  ADReal r = 0;
  switch (type)
  {
    
    case Moose::Element:
      r= -_lstandard[_qp] * 96485.33/8.314/293.15* (_u[_qp]-_neighbor_value[_qp]-_Eeq[_qp]) *_test[_i][_qp] * _cl[_qp];
      break;

    
    case Moose::Neighbor:
      r = _lstandard[_qp] * 96485.33/8.314/293.15* (_u[_qp]-_neighbor_value[_qp]-_Eeq[_qp])*_test_neighbor[_i][_qp] * _cl[_qp];
      break;
  }
  return r;
}
```
BVFLUX:
```
#include "BVFLUX.h"

registerMooseObject("batteryApp", BVFLUX);

InputParameters
BVFLUX::validParams()
{
  InputParameters params = ADInterfaceKernel::validParams();
  params.addCoupledVar("phis","solid potential");
  params.addCoupledVar("phil","liquid potential");
  params.addParam<MaterialPropertyName>("i0","name of i0");
  params.addParam<MaterialPropertyName>("Eeq","name of Eeq");
  params.addClassDescription("Implements a reaction to establish ReactionRate=k_f*u-k_b*v "
                             "at interface.");
  return params;
}

BVFLUX::BVFLUX(const InputParameters & parameters)
  : ADInterfaceKernel(parameters),
  _phis(coupledValue("phis")),
  _phil(coupledValue("phil")),
  _lstandard(getMaterialProperty<Real>("i0")),
  _Eeq(getMaterialProperty<Real>("Eeq"))
{
}

ADReal
BVFLUX::computeQpResidual(Moose::DGResidualType type)
{
  ADReal r = 0;
  switch (type)
  {
    case Moose::Element:
      r=-_lstandard[_qp] * 1/8.314/293.15* (_phis[_qp]-_phil[_qp]-_Eeq[_qp])*_test[_i][_qp]*_u[_qp];
      break;
  }
  return r;
}
```
My input file:
```
[Mesh]
    type = FileMesh
    file = mesh_realsize.inp
[]

[Variables]
    [phis]
        block='solid_TRI3'
    []
    [phil]
        block='liquid_TRI3'
    []
    [cl]
        block='liquid_TRI3'
    []
[]

[Materials]
    [D_solid]
        type=GenericConstantMaterial
        prop_names='sigmas sigmal Dl Eeq i0'
        prop_values='1 1 1e-10 1 0.1'
    []
[]

[Kernels]

    [ohm_s]
        block='solid_TRI3'
        type=MatDiffusion
        variable=phis
        diffusivity='sigmas'
    []

    [ohm_l]
        block='liquid_TRI3'
        type=MatDiffusion
        variable=phil
        diffusivity='sigmal'
    []

    [diff_l]
        block='liquid_TRI3'
        type=MatDiffusion
        variable=cl 
        diffusivity='Dl'
    []
    [time_l]
        block='liquid_TRI3'
        type=TimeDerivative
        variable=cl 
    []
[]

[ICs]
    [phis_ic]
        type=ConstantIC
        variable=phis
        value=0
        block='solid_TRI3'
    []
    [phil_ic]
        type=ConstantIC
        variable=phil
        value=0
        block='liquid_TRI3'
    []
    [cl_ic]
        type=ConstantIC
        variable=cl 
        value=1000
        block='liquid_TRI3'
    []
[]

[BCs]
    [phis_neumann]
        boundary='bottom'
        type=ADNeumannBC
        variable=phis
        value=-1
    []

    [phil_dirichlet]
        boundary='top'
        type=ADDirichletBC
        variable=phil
        value=0
    []

    [cl_dirichlet]
        boundary='top'
        type=ADDirichletBC
        variable=cl
        value=1000
    []
[]

[InterfaceKernels]
    [BV]
        type=BV 
        variable=phis
        neighbor_var=phil
        cl=cl
        i0='i0'
        Eeq='Eeq'
        boundary='interface'
    []
    [BVFLUX]
        type=BVFLUX
        variable=cl 
        neighbor_var=phis //here I just put phis as a place holder because InterfaceKernel demand a neighbor_var.
        phis=phis
        phil=phil
        i0='i0'
        Eeq='Eeq'
        boundary='interface'
    []
[]


[Executioner]
    type=Transient
    num_steps=10
    dt=10
    automatic_scaling=true
    solve_type=NEWTON
[]

[Outputs]
    exodus=true
[]
```

**Comment by GiudGiud** [comment]
Hello

What do you mean by cannot be calculated? Did you get an error?

One thing to be careful about when using interface kernels is that the sideset, here 'interface' is consistently oriented

Guillaume

**Reply by Chaoyue-Liu** [reply]
Hello,
The model cannot converge at the first step. I got the log like this and I have to kill the calculation:
![image](https://user-images.githubusercontent.com/95432912/195139788-2db53174-be7c-4f0b-9851-cb152b6b10d6.png)

**Comment by hugary1995** [comment, marked as answer]
A known issue: if the variable isn't defined on both sides of the interface, the interface kernel won't be calculated. That happens silently so your simulation could still run without warning or error.

So... Is your solid concentration block restricted?

**Reply by hugary1995** [reply]
Hmm, I see that you defined a dummy neighbor var. Idk what's going on then...

**Reply by Chaoyue-Liu** [reply]
Yes, phis in BVFLUX is just place holder, it's not actually used in the residual calculation.

**Reply by hugary1995** [reply]
You phil has a Dirichlet BC so that is okay. The phis only has a Neumann BC at the bottom, so if the butler-volmer condition isn't active, there will indeed be a null space, hence the convergence log you saw (i.e. linear solve stagnates).

Check if removing the BV interface changes the residual and/or the convergence log. If nothing changes, then we can look into fixing the BV kernel.

**Reply by hugary1995** [reply]
Alternatively, reduce the problem size and run with `-pc_type svd -pc_svd_monitor` to see if any of the singular values is close to zero.

**Reply by Chaoyue-Liu** [reply]
Thank you hugary. I did a test which doesn't involve equation 2) (BVFLUX kernel), only equation 1) (BV kernel) is active for bridging phis and phil. The cl domain is changed to covering both liquid_TRI3 and solid_TRI3, with dirichlet on top and neumann on bottom, just to provide cl value to equation 1). The calculation converges fine and the result is correct. Once I involve equation 2) the calculation is not able to proceed. I'm consistently comparing the result with COMSOL with the same equation setup. In COMSOL it's able to converge with equation 2) so I guess the equations are fine. I hope this test replies your first question.

My geometry size is 50µm, I think it's small enough already? And how do I use `-pc_type svd -pc_svd_monitor` exactly? I'm not expert on this sorry.

**Reply by hugary1995** [reply]
ok, so it seems like the mass balance is affecting the convergence here. But that shouldn't happen as there is a time derivative in cl. Feel free to take a look at what I did here https://github.com/hugary1995/eel/blob/main/test/tests/chemical-electrical/charging.i and work your way back.

`-pc_type svd -pc_svd_monitor` are petsc options, you either pass it as cli args, or specify them in the input file as
```python
[Executioner]
  petsc_options = '-pc_svd_monitor'
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'svd'
[]
```

**Reply by Chaoyue-Liu** [reply]
Thank you hugary! I'll look into it.

**Reply by GiudGiud** [reply]
Any update on this?

**Reply by Chaoyue-Liu** [reply]
Hi Guillaume, the problem is solved. It's because I used CoupledValue instead of CoupledNeighborValue for the variable in neighbor element. Also the relative tolerance need ajustment. Now the kernel works fine. Thank you!

**Comment by Chaoyue-Liu** [comment]
For the convenience of discussion, here is a repository I created with the files I use for your reference: 
https://github.com/Chaoyue-Liu/battery.git
